File: auto/src/domain/valueObjects/Genres.ts
import { __brand, Branded } from './BrandedTypes';

const genres: string[] = [
  'Fantasy',
  'Science Fiction',
  'Horror',
  'Space Opera',
  'Mystery',
  'Cyberpunk',
  'Romance',
  'Steampunk',
  'High Fantasy',
  'Low Fantasy',
  'Cthulhu Mythos',
  'LitRPG',
  'GameLit',
  'Hard Sci-Fi',
  'Soft Sci-Fi',
  'Military Sci-Fi',
  'Superhero',
  'Post-Apocalyptic',
] as const; // Marking it as a const makes it a tuple so we can use it as a type. - Even though we declare it as a const at the start we need this to correctly infer the tuple type property.

type Genre = (typeof genres)[number];

class Genres {
  private static readonly genres = genres;

  static getRandomGenre(): Genre {
    return this.genres[Math.floor(Math.random() * this.genres.length)];
  }
}

export { Genre, Genres };

File: auto/domain/valueObjects/genres.go
package valueObjects

import "math/rand"

type Genre string

const (
	Fantasy         Genre = "Fantasy"
	ScienceFiction  Genre = "Science Fiction"
	Horror          Genre = "Horror"
	SpaceOpera      Genre = "Space Opera"
	Mystery         Genre = "Mystery"
	Cyberpunk       Genre = "Cyberpunk"
	Romance         Genre = "Romance"
	Steampunk       Genre = "Steampunk"
	HighFantasy     Genre = "High Fantasy"
	LowFantasy      Genre = "Low Fantasy"
	CthulhuMythos   Genre = "Cthulhu Mythos"
	LitRPG          Genre = "LitRPG"
	GameLit         Genre = "GameLit"
	HardSciFi       Genre = "Hard Sci-Fi"
	SoftSciFi       Genre = "Soft Sci-Fi"
	MilitarySciFi   Genre = "Military Sci-Fi"
	Superhero       Genre = "Superhero"
	PostApocalyptic Genre = "Post-Apocalyptic"
)

var genres = []Genre{
	Fantasy, ScienceFiction, Horror, SpaceOpera, Mystery,
	Cyberpunk, Romance, Steampunk, HighFantasy, LowFantasy,
	CthulhuMythos, LitRPG, GameLit, HardSciFi, SoftSciFi,
	MilitarySciFi, Superhero, PostApocalyptic,
}

func GetRandomGenre() Genre {
	return genres[rand.Intn(len(genres))]
}
